amplipi/app.py

```diff
@@ -897,60 +898,78 @@ def get_ip_info(iface: str = 'eth0') -> Tuple[Optional[str], Optional[str]]:
     return None, None
 
 
-def advertise_service(port, que: Queue):
+def advertise_service(port, event: SyncEvent):
   """ Advertise the AmpliPi api via zeroconf, can be verified with 'avahi-browse -ar'
       Expected to be run as a seperate process, eg:
 
-          q = Queue()
-          ad = Process(target=amplipi.app.advertise_service, args=(5000, q))
+          event = multiprocessing.Event()
+          ad = Process(target=amplipi.app.advertise_service, args=(5000, event))
           ad.start()
           ...
-          q.put('done')
+          event.set()
           ad.join()
+      NOTE: multiprocessing.Event() is a function that returns a multiprocessing.synchronize.Event type
+      Here the type is aliased to SyncEvent
   """
+  def ok():
+    """ Was a stop requested by the parent process? """
+    return not event.is_set()
+
+  while ok():
+    try:
+      _advertise_service(port, ok)
+    except Exception as exc:
+      if not 'change registration' in str(exc):
+        print(f'Failed to advertise AmpliPi service, retrying in 30s: {exc}')
+        # delay for a bit after a failure
+        delay_ct = 300
+        while ok() and delay_ct > 0:
+          sleep(0.1)
+          delay_ct -= 1
+
+
+def _find_best_iface(ifaces: List[str], ok: Callable) -> Union[Tuple[str, str, str], Tuple[None, None, None]]:
+  # Try to find the best interface to advertise on,
+  # retrying in case DHCP resolution is delayed
+  ip_addr, mac_addr = None, None
+  retry_count = 5
+  while ok() and not (ip_addr and mac_addr) and retry_count > 0:
+    for iface in ifaces:
+      ip_addr, mac_addr = get_ip_info(iface)
+      if ip_addr and mac_addr:
+        return ip_addr, mac_addr, iface
```

Comment:
Calling the first interface with an IP address the 'best' interface is probably valid in most cases, but this still fails on my computer. The first Ethernet port, call it eth0, has a local address with no internet, so it's not the default gateway. The second Ethernet port, call it eth1, is the default gateway. `test_rest.py:test_zeroconf()` advertises on ni.gateways()['default'], so eth1. AmpliPi advertises on the alphabetically first eth0, so the test fails.

A default gateway isn't guaranteed to exist, for example if a computer and AmpliPi are connected link-locally. But clearly selecting the first IP address isn't perfect either. I'd suggest checking for a default gateway and if present using that. If not present, then fall back to the first IP address method here.

In the case where someone has plugged in a USB network adapter into an AmpliPi (most likely WiFi but could be Ethernet) and they both have valid routes to a real network, I'm not sure we can do anything so I guess I'm not worried about handling that. Unless we can do multiple advertisements if there are multiple valid interfaces with IPs?